NEW DataObject related objects service

emteknetnz opened the pull request:
Issue https://github.com/silverstripe/silverstripe-admin/issues/1098

Blocks https://github.com/silverstripe/silverstripe-admin/pull/1139

I used a service class for this because DataObject is the logical place for this code to live, though that file is already over 4,000 lines long :-)

maxime-rainville reviewed (changes requested):
Good work on improving the logic. All my pre-existing scenario work fine now.

I did notice that blocks from deleted page still show up on the table. They also seem to be very slow to retrieve
![image](https://user-images.githubusercontent.com/1168676/96207538-f350cf00-0fc7-11eb-8bd3-889ac54279ad.png)

maxime-rainville commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,332 @@
+            $tableName = '';
+            $candidateClass = $class;
+            // Using a for loop rather than a while loop just to be safe
+            for ($i = 0; $i < 25 && $candidateClass; $i++) {
```
Are we expecting some sort of recursion here? We ought to have a good understanding of our exit condition here and make sure it's always met at some stage rather than rely on an arbitrary loop limit.

emteknetnz replied:
Changed it to a while loop

maxime-rainville commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,332 @@
+     * @return ArrayList
+     * @throws \ReflectionException
+     */
+    public function findAll(DataObject $record, array $excludedClasses = []): ArrayList
```
```suggestion
    public function findAll(DataObject $record, array $excludedClasses = []): SSList
```

maxime-rainville replied:
I still think there's a lot of opportunity to split this method into smaller more manageable chunks.

emteknetnz replied:
I refactored the whole thing, lots of private methods now

maxime-rainville commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,332 @@
+
+        // Loop has_one relationships on the DataObject we're getting usage for
+        // e.g. File.has_one = Page, Page.has_many = File
+        $class = get_class($record);
```
Your record already has all the methods needed to retrieve relation data. Reading it from the config requires the duplication of logic.

emteknetnz replied:
Updated

emteknetnz commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,332 @@
+use SilverStripe\ORM\DataObjectSchema;
+use SilverStripe\ORM\Queries\SQLSelect;
+
+class DataObjectRelatedService
```
I'm not sure why we need define at interface for the service?  We've made it injectable so people can replace it with their own implementation if they want to. An interface seems like a bit of over abstraction?

robbieaverill replied:
I think writing interfaces for your service classes and using those via Injector to get them is good practice actually. It makes the dependencies in our systems more compliant with the dependency inversion principle (D) of SOLID

emteknetnz replied:
So, literally just do this?

```php

interface DataObjectRelatedServiceInterface
{
    public function findAll(DataObject $record, array $excludedClasses = []): SS_List
}

class DataObjectRelatedService implements DataObjectRelatedServiceInterface
{
    use Injectable;

    public function findAll(DataObject $record, array $excludedClasses = []): SS_List
    {
        ...
}
```

?

emteknetnz replied:
Updated

emteknetnz commented on src/ORM/DataObject.php:
```diff
@@ -4348,4 +4349,15 @@ private function getUniqueKeyComponents(): array
+     * @return ArrayList
+     * @throws \ReflectionException
+     */
+    public function findAllRelatedObjects(array $excludedClasses = []): ArrayList
```
Updated

emteknetnz commented on src/ORM/DataObject.php:
```diff
@@ -4348,4 +4349,15 @@ private function getUniqueKeyComponents(): array
+     */
+    public function findAllRelatedObjects(array $excludedClasses = []): ArrayList
+    {
+        $service = new DataObjectRelatedService();
```
Updated

emteknetnz commented on src/ORM/DataObject.php:
```diff
@@ -4348,4 +4349,15 @@ private function getUniqueKeyComponents(): array
     }
+
+    /**
+     * @param array $excludedClasses
```
Updated

emteknetnz commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,332 @@
+            }
+
+            // Loop has_one and many_many within the config data for the class
+            foreach ($relationTypes as $relationType) {
```
Updated

emteknetnz commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,332 @@
+
+            // Loop has_one and many_many within the config data for the class
+            foreach ($relationTypes as $relationType) {
+                $configData = Config::forClass($class)->get($relationType);
```
Updated

maxime-rainville commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,437 @@
+    }
+
+    /**
+     * Query the database to retrieve has-one results
+     *
+     * @param DataObject $record - The DataObject whose usage data is being retrieved, usually a File
+     * @param string $class - example: My\App\SomePageType
+     * @param string $component - example: 'SomeFiles' - My\App\SomePageType::SomeFiles()
+     * @return Query|null
```
```suggestion
     * Query the database to retrieve has-one results.
     *
     * @param DataObject $record - The DataObject whose usage data is being retrieved, usually a File
     * @param string $class Name of class with the relation to record
     * @param string $component Name of relation to `$record` on `$class`
     * @return Query|null
```

emteknetnz replied:
Updated

maxime-rainville commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,437 @@
+     * @return ArrayList
+     * @throws \ReflectionException
+     */
+    public function findAll(DataObject $record, array $excludedClasses = []): SS_List
```
How do we handle records that haven't been saved to the database, singletons or archived records?

emteknetnz replied:
>> haven't been saved to the database

These will not be included in the results.  That seems like a real edge case.  In the context of the used on table piece of work this is being done around it's irrelevant.

>> singletons

I'm not even sure what this has to do with related data

>> archived records

Not included in the results.  This service largely bypasses the ORM for performance and reads from draft tables only.

maxime-rainville replied:
People might use `findAll` for other purposes.

I'm fine with it returning nothing or throwing a descriptive exception if you give it an unsaved record or other non-sense. Just don't want it to keep trying to find data for a non-existent record and then throw some cryptic error when it tries to do a query with a non-ID.

emteknetnz replied:
I've updated it to return an empty ArrayList when `!$record->exists()`

maxime-rainville commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,437 @@
+    }
+
+    /**
+     * @param array $classIDs
```
```suggestion
     * @param int[] $classIDs
```

emteknetnz replied:
Updated

maxime-rainville commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,437 @@
+    }
+
+    /**
+     * Update the $classIDs array with the relationship IDs from database $results
+     *
+     * @param array $classIDs
```
```suggestion
     * Update the `$classIDs` array with the relationship IDs from database `$results`.
     *
     * @param int[] $classIDs
```

emteknetnz replied:
$classIDs isn't an int[], it's an assoc_array where $key = string (ClassName) and $value = int (ID)

maxime-rainville commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,437 @@
+            return;
+        }
+        foreach ($results as $row) {
+            if (count(array_keys($row)) == 2 && isset($row['ParentClass']) && isset($row['ParentID'])) {
```
Presumably that's for polymorphic relations? Maybe add a comment to make this explicit.

```suggestion
            if (count(array_keys($row)) === 2 && isset($row['ParentClass']) && isset($row['ParentID'])) {
```

emteknetnz replied:
Updated

maxime-rainville reviewed (changes requested):
Much better. Left some extra comments, but we're almost there I think.

maxime-rainville reviewed (changes requested):
Tested this against polymorphic relations, many-many-through relations and self referential relations.

Nothing blew up in my face.

Has one from File to something else and many-many-through didn't show on the used on table ... which may be the intended behavior.

Self-referential relation show up twice in the table, not clear what the expected behaviour is here.

Orphan content blocks are still incredibly slow for some reason that I didn't investigate.

emteknetnz commented on src/ORM/DataObject.php:
```diff
@@ -4348,4 +4350,20 @@ private function getUniqueKeyComponents(): array
+     * This method is called on a File.  The MyPage model $has_one File.  There is a Page record that has
+     * a FileID = $this->ID. This SS_List returned by this method will include that Page instance.
+     *
+     * @param array $excludedClasses
```
Updated

emteknetnz commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,437 @@
+     * Find all DataObject instances that have a linked relationship with $record
+     *
+     * @param DataObject $record
+     * @param array $excludedClasses
```
Updated

emteknetnz commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,437 @@
+     * Loop has_one relationships on the DataObject we're getting usage for
+     * e.g. File.has_one = Page, Page.has_many = File
+     *
+     * @param array $classIDs
```
Updated

emteknetnz commented on src/ORM/Related/DataObjectRelatedService.php:
```diff
@@ -0,0 +1,437 @@
+    }
+
+    /**
+     * @param array $classIDs
```
Updated

maxime-rainville commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,433 @@
+ * - $component: 'MyFile'
+ * - $componentClassName: SilverStripe\Assets\File::class
+ */
+class StandardRelatedDataService
```
```suggestion
class StandardRelatedDataService implements RelatedDataService
```

maxime-rainville commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,433 @@
+ * - $component: 'MyFile'
+ * - $componentClassName: SilverStripe\Assets\File::class
+ */
+class StandardRelatedDataService
```
```suggestion
class StandardRelatedDataService
```

emteknetnz replied:
Updated

maxime-rainville commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,433 @@
+        return SQLSelect::create(
+            '"ID"',
+            sprintf('"%s"', $tableName),
+            sprintf('"%s" = %s', $componentIDField, $record->ID)
```
```suggestion
            sprintf('"%s" = %u', $componentIDField, $record->ID)
```

emteknetnz replied:
Updated

maxime-rainville reviewed (changes requested):
Applying the interface to your service is the only must fix thing.

You'll probably need to update the doc to explain to dev how to customise the used on table. Either do it in this PR or do a separate DOC only PR

maxime-rainville commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,433 @@
+        }
+        $usesThroughTable = $data['join'] != $joinTableName;
+
+        $parentField = rtrim($data['parentField'], 'ID') . 'ID';
```
That's still not going to work if you have a column called something like `MyColumnIDIDIDIDID'

emteknetnz replied:
Updated to use preg_replace

emteknetnz commented:
@maxime-rainville 

>> Self-referential relation show up twice in the table, not clear what the expected behaviour is here.

I've added some unit testing for has_one and many_many and updated code.  Now it returns 1 instead 2 results when a DataObject refers to itself.

>> Orphan content blocks are still incredibly slow for some reason that I didn't investigate.

In the context of the `StandardRelatedDataService` it should still returns the orphaned content blocks

In the context of the used on table, I think they could be pruned [here](https://github.com/dnadesign/silverstripe-elemental/pull/827/files#diff-5704678f677ed41a8453cbe886816a8276c95a6376465b4acdead58cdfd35796R54)

maxime-rainville commented:
I've updated my [GIST with the tasks to create all kind of relations](https://gist.github.com/maxime-rainville/26785a996331af41298c71eaee7e52b4).

The only relation types that don't appear to be working right now is the `manyManyThrough`.

> I've added some unit testing for has_one and many_many and updated code. Now it returns 1 instead 2 results when a DataObject refers to itself.

Logically, we should use the same logic when deciding if we need to warn the user when they delete a file. This means you'll alaways get a warning when deleting a self-referential File. You also end up with an entry in the Used on table that link to itself. 

Don't really have a great answer here. Probably should ask PUX about this.

>> Orphan content blocks are still incredibly slow for some reason that I didn't investigate.
>
> In the context of the StandardRelatedDataService it should still returned the orphaned content blocks because DataObject isn't aware of how Content Block -> Page relations work

The slowness is probably caused by the logic to build the ancestor tree. This shouldn't old off this specific PR.

maxime-rainville reviewed (changes requested):
I'm still not seeing the many many through when using my CreateRelated task.

emteknetnz commented:
@maxime-rainville 

It's supposed to look like this yes?

![image](https://user-images.githubusercontent.com/4809037/97259951-d1d1cc00-1880-11eb-9ee0-3fdf586196f2.png)

Have you pulled the latest?  `StandardRelatedDataService` should have this code in there which fixed this issue:

https://github.com/silverstripe/silverstripe-framework/compare/536f740fea5e847be1fbc8312f59b6986418c86d..8cbdabb

sminnee reviewed (commented):
A few comments suggesting we should think carefully before adding a new public API

maxime-rainville reviewed (approved):
Many-many-through are working.

Regarding Sam's concern, I'm fine with marking the new API with `@internal`. That would give us more flexibility to patch it in 4.8 if we end up messing something up.

emteknetnz commented on src/ORM/DataObject.php:
```diff
@@ -4348,4 +4349,19 @@ private function getUniqueKeyComponents(): array
     {
         return $this->extend('cacheKeyComponent');
     }
+
```
It's not strictly required, and yes we could just put this code in admin UsedOnTable.php, though it's kind of nice having it directly on DataObject.

Having it on DataObject would make it much more 'findable' to other developers and encourage them to use it because they'll get the typehint in their IDE

Would it work if we also put `@internal` on this function for now, along with the `@internal` on the interface?

maxime-rainville commented:
@sminnee's feeback has been addressed. Marking the new API as `@internal` gives us flexibility to change them later on if we find limitation.

tractorcow commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,507 @@
+
+            // Polymorphic
+            // $record->ParentClass will return null if the column doesn't exist
+            if ($componentIDField === 'ParentID' && $record->ParentClass) {
```
I'm a bit late to this party, but if you want to detect polymorphic has_ones, you should be checking `$componentClass === DataObject::class`.

This code has a incorrect baked-in assumption that ONLY relations named ParentID can be polymorphic, when any relation can be polymorphic at all. E.g. `RelatedID, RelatedClass`.

The code could even have a `getParentClass()` method that would trick this code into thinking there's a polymorphic relation too.

tractorcow commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,507 @@
+     * @return string
+     */
+    private function findTableNameContainingComponentIDField(string $class, string $componentIDField): string
+    {
```
This code is duplicating DataObjectSchema::tableForField()

tractorcow commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,507 @@
+        // ParentID: int, ParentClass: enum('File::class, SiteTree::class, ElementContent::class, ...')
+        if ($usesThroughTable) {
+            $dbFields = $this->dataObjectSchema->databaseFields($data['join']);
+            if ($parentField === 'ParentID' && isset($dbFields['ParentClass'])) {
```
Again, assuming that ParentID is specifically the only polymorphic field? Same critique as above.

tractorcow commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,507 @@
+        return SQLSelect::create(
+            sprintf('"' . implode('", "', $selectFields) . '"'),
+            sprintf('"%s"', $joinTableName),
+            sprintf('"%s" = %u', $whereField, $record->ID)
```
You should be parameterising $record->ID to improve performance; DB layer can re-use prepared queries without baking in a hard-coded ID value.

tractorcow commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,507 @@
+        }
+
+        $this->config = Config::inst()->getAll();
+        $this->dataObjectSchema = DataObjectSchema::create();
```
You should be using ::singleton() not ::create().

Your issue with test classes is that sapphiretest will flush the cache of dataobject singleton, not ALL instances that are created. :)

Or just DataObject::getSchema() instead of storing it in a protected property.

tractorcow commented on src/ORM/RelatedData/StandardRelatedDataService.php:
```diff
@@ -0,0 +1,507 @@
+            return ArrayList::create();
+        }
+
+        $this->config = Config::inst()->getAll();
```
Sorry for sounding alarmist, but storing the ENTIRE config in a variable is a very bad way to "discover" a lists of all DataObjects. You should be using some other mechanism for discovering dataobjects, such as the class manifest.

Try `ClassInfo::subclassesFor(DataObject::class, false)` instead.

tractorcow commented:
Hi, it's me from the dead. :)

I came back to this PR after discovering some wierdness with 4.7. (I'm still debugging). There are a few QA concerns with this code that deserve someone giving this a re-review. I'll have to come back to this when I have time though. :)

emteknetnz commented:
Hi! Good to see you again :-)

> after discovering some wierdness with 4.7

Would you be able to raise a new issue on this repo that includes the word 'DataRelatedService' and describe the weirdness you're experiencing.  Having a new issue will make it more discoverable from others in the community to add any weirdness they are seeing on their own projects

tractorcow commented:
Sure,

It was actually on a discussion in slack that I was looking into on behalf of a user there.

https://silverstripe-users.slack.com/archives/C36U3BDHN/p1617127694283700?thread_ts=1616424417.130700&cid=C36U3BDHN

I might raise a PR instead of an issue, depending on how my afternoon goes.

emteknetnz commented:
Great, thanks for letting us know.  I've raised an issue for this https://github.com/silverstripe/silverstripe-framework/issues/9902
